## Analysis of Titanic Survival Rates
The Titanic disaster is one of the most well-known maritime tragedies in history. This project explores the classic dataset of Titanic passengers to uncover factors that influenced a passenger's likelihood of survival. By cleaning, exploring, and analyzing the data, we aim to answer questions such as: Did gender, wealth, or family size play a role in who survived?

This analysis uses the pandas library for data manipulation and seaborn for loading the dataset.

In [42]:
import pandas as pd
import seaborn as sns

## Data Loading and Initial Inspection
At first, we load the dataset directly from the Seaborn library and perform an initial inspection to understand its structure, data types, and identify any immediate issues like missing values.

### Loading the Dataset

In [43]:
df = sns.load_dataset('titanic')

### Viewing the First Few Rows of the Dataset

In [44]:
print('The first 10 rows of the Dataset are as follows:')
df.head(10)

The first 10 rows of the Dataset are as follows:


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
5,0,3,male,NaN,0,0,8.4583,Q,Third,man,True,NaN,Queenstown,no,True
6,0,1,male,54.0,0,0,51.8625,S,First,man,True,E,Southampton,no,True
7,0,3,male,2.0,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
8,1,3,female,27.0,0,2,11.1333,S,Third,woman,False,NaN,Southampton,yes,False
9,1,2,female,14.0,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False


### What is the size of the Dataset?

In [45]:
print('The size of the dataset is :',df.shape)

The size of the dataset is : (891, 15)


### What type of Data Types are Present in the Dataset?

In [46]:
print('The dataset has the following Data Types : ') 
df.info()

The dataset has the following Data Types : 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


## Checking for Null Values

In [47]:
print('Checking for Null Values in Different Columns :')
df.isnull().sum()

Checking for Null Values in Different Columns :


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

## Observations

The age column has 177 missing values. We will fill these with the median age, which is more robust to outliers than the mean.

The embarked and embark_town columns have 2 missing values. We will fill them with the mode (the most common port).

The deck column is missing too many values (688 out of 891) to be useful, so we will drop it.

The data type of Sex Column should be changed to catgeory instead of Object.

### Substituting the missing 'age' values with the median age

In [48]:
median_age = df['age'].median()
df['age'] = df['age'].fillna(median_age)

### Then we fill the missing 'embarked' and 'embark_town' values with the mode

In [49]:
mode_embarked = df['embarked'].mode()[0]
mode_embark_town = df['embark_town'].mode()[0]
df['embarked'] = df['embarked'].fillna(mode_embarked)
df['embark_town'] = df['embark_town'].fillna(mode_embark_town)

### Since the Deck Column has 688 missing values out of 891 we Drop It

In [50]:
df.drop(columns=['deck'], inplace = True)

### Now we convert the 'sex' column to from Object to Category type for efficiency

In [51]:
df['sex'] = df['sex'].astype('category')

### Verifying if all the missing values are  now gone

In [52]:
df.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

### Now we obtain some insights from the Dataset

To better analyze survival patterns, we can create new columns (features) from the existing data. We will create columns to determine a passenger's family size and whether they were traveling alone.

### Addition of family_size column

In [54]:
df['family_size'] = df['sibsp'] + df['parch'] + 1

### Addition of a Column In order to check if the person is living alone

In [56]:
df['is_alone'] = (df['family_size'] == 1)

### Now we see the first few rows of the dataset in order to see the newly added rows

In [57]:
df.head(8)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone,family_size,is_alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False,2,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False,2,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True,1,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False,2,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True,1,True
5,0,3,male,28.0,0,0,8.4583,Q,Third,man,True,Queenstown,no,True,1,True
6,0,1,male,54.0,0,0,51.8625,S,First,man,True,Southampton,no,True,1,True
7,0,3,male,2.0,3,1,21.0750,S,Third,child,False,Southampton,no,False,5,False


### Exploratory Data Analysis (EDA)

Now that our data is clean, we can answer our key questions by grouping and aggregating the data.

### Q1: What was the overall survival rate?

In [63]:
survival_rate = df['survived'].mean()

In [67]:
print(f'The overall survival rate was : {survival_rate:.2%}')

The overall survival rate was : 38.38%


### Q2: Was the overall survival rate influenced by Gender?

In [76]:
gen_surv_rate = df.groupby('sex', observed = True)['survived'].mean()

In [81]:
gen_surv_rate.to_frame().style.format('{:.2%}')

,survived
sex,
female,74.20%
male,18.89%


Observation - Female passengers had a much higher survival rate (74%) compared to male passengers (19%).

### Q3: Did the class in which passengers travelled influence survival?

In [83]:
class_surv = df.groupby('pclass')['survived'].mean()

In [84]:
class_surv.to_frame().style.format('{:.2%}')

,survived
pclass,
1,62.96%
2,47.28%
3,24.24%


Observation - There is a clear correlation between wealth and survival. First-class passengers had the highest survival rate (63%), while third-class passengers had the lowest (24%).

### Q4. Was survival of an Individual Impacted by travelling alone?

In [86]:
alone_surv = df.groupby('is_alone')['survived'].mean()

In [88]:
alone_surv.to_frame().style.format('{:.2%}')

,survived
is_alone,
False,50.56%
True,30.35%


Observation - Passengers who were with family had a higher survival rate (50.6%) than those who were traveling alone (30.4%).

### Conclusion

This analysis of the Titanic dataset reveals several clear factors that influenced survival. The data strongly suggests that females, passengers in higher classes, and those traveling with family had a significantly greater chance of surviving the disaster. This project demonstrates a complete data analysis workflow: loading, inspecting, cleaning, feature engineering, and drawing conclusions from the data using pandas.